kotlinx-coroutines-core/common/src/channels/BufferedChannel.kt

```diff
@@ -0,0 +1,3087 @@
+    ): R {
+        // Read the segment reference before the counter increment;
+        // it is crucial to be able to find the required segment later.
+        var segment = sendSegment.value
+        while (true) {
+            // Atomically increment the `senders` counter and obtain the
+            // value right before the increment along with the close status.
+            val sendersAndCloseStatusCur = sendersAndCloseStatus.getAndIncrement()
+            val s = sendersAndCloseStatusCur.sendersCounter
+            // Is this channel already closed? Keep the information.
+            val closed = sendersAndCloseStatusCur.isClosedForSend0
+            // Count the required segment id and the cell index in it.
+            val id = s / SEGMENT_SIZE
+            val i = (s % SEGMENT_SIZE).toInt()
+            // Try to find the required segment if the initially obtained
+            // one (in the beginning of this function) has lower id.
+            if (segment.id != id) {
+                // Find the required segment.
+                segment = findSegmentSend(id, segment) ?:
+                    // The required segment has not been found.
+                    // Finish immediately if this channel is closed,
+                    // restarting the operation otherwise.
+                    // In the latter case, the required segment was full
+                    // of interrupted waiters and, therefore, removed
+                    // physically to avoid memory leaks.
+                    if (closed) {
+                        return onClosed()
+                    } else continue
+            }
+            // Update the cell according to the algorithm. Importantly, when
+            // the channel is already closed, storing a waiter is illegal, so
+            // the algorithm stores the `INTERRUPTED_SEND` token in this case.
+            when(updateCellSend(segment, i, element, s, waiter, closed)) {
+                RESULT_RENDEZVOUS -> {
+                    // A rendezvous with a receiver has happened.
+                    // The previous segments are no longer needed
+                    // for the upcoming requests, so the algorithm
+                    // resets the link to the previous segment.
+                    segment.cleanPrev()
+                    return onRendezvousOrBuffered()
+                }
+                RESULT_BUFFERED -> {
+                    // The element has been buffered.
+                    return onRendezvousOrBuffered()
+                }
+                RESULT_SUSPEND -> {
+                    // The operation has decided to suspend and installed the
+                    // specified waiter. If the channel was already closed,
+                    // and the `INTERRUPTED_SEND` token has been installed as a waiter,
+                    // this request finishes with the `onClosed()` action.
+                    if (closed) {
+                        segment.onSlotCleaned()
```

Comment:
It may be reachable. Consider a situation when a channel is marked as closed, but the cell is still empty. In this case, the operation installs `INTERRUPTED_SEND` in the cell. I missed this part of the algorithm when documenting the code -- will fix that.